src/lib.rs

```diff
@@ -0,0 +1,489 @@
+    Tracing(JitLocation),       // Recording a trace from the specified location.
+    RunningTrace(JitLocation),  // Running a trace starting from the specified location.
+}
+
+/// A machine code trace for a location.
+#[derive(Debug, PartialEq, Eq)]
+enum JitTrace {
+    TracePending,       // Currently being traced or compiled.
+    FinalisedTrace,     // A finished trace with executable code.
+}
+
+fn debug_thread_name<'a>() -> String {
+    let thr = std::thread::current();
+    let sref = match thr.name() {
+        Some(n) => n,
+        None => "no_name",
+    };
+    String::from(sref)
+}
+
+/// JIT Control Point.
+///
+/// The interpreter will need to call this periodically to inform the meta-tracer where in the
+/// user-program we are. For example, in a bytecode interpreter, the JIT control point will be
+/// called immediately inside the bytecode dispatch loop.
+///
+/// This function is passed a JitLocation, which should uniquely identify the position in the
+/// user-program (usually this includes at least a program counter value). The JIT will use this
+/// information for profiling, trace collection, and trace execution.
+///
+/// The control point returns a JitTransition indicating the work the JIT undertook.
+pub fn jit_control_point(loc: JitLocation) -> JitTransition {
+    debug!("JIT Control: thread={:?}: loc={:?}, hot_count={}",
+           debug_thread_name(), loc, hot_count(loc));
+
+    let ret;
+    match JitThreadContext::current_phase() {
+        JitPhase::Interpreting => {
+            let ran = try_exec_trace(loc);
+            if ran {
+                ret = JitTransition::ExecutedTrace;
+            } else {
+                // There is no trace for this location (yet), so we just increment the hot count.
+                let ct = incr_hot_count(loc);
+                if ct == HOT_THRESHOLD && !trace_pending(loc) {
+                    // Location becomes hot.
+                    set_trace(loc, JitTrace::TracePending);
+                    // XXX start Intel PT.
+                    JitThreadContext::phase_transition(JitPhase::Tracing(loc));
+                    ret = JitTransition::StartedTracing;
+                } else {
+                    ret = JitTransition::IncrementedHotCount;
+                }
+            }
+        },
+        JitPhase::Tracing(start_loc) => {
+            // If we have traced one iteration of the loop, stop tracing.
+            if loc == start_loc {
+                // XXX stop Intel PT.
+                JitThreadContext::compile_trace(loc);
+                JitThreadContext::phase_transition(JitPhase::Interpreting);
+                ret = JitTransition::FinishedTracing;
+            } else {
+                // Still tracing.
+                ret = JitTransition::ContinuedTracing;
+            }
+        },
+        JitPhase::RunningTrace(_) => {
+            // The JIT control point should not be called from within a trace.
+            unreachable!();
```

Comment:
I think we need to be very cautious about `unreachable`. As I read it, it's purely an optimisation hint. If the assumptions underlying that turn out to be wrong, the system will then do weird things. I think this should probably be a `panic`?

Reply:
It's not just an optimisation hint, it's a panic too:
https://play.rust-lang.org/?gist=dbd01a2666c0c8de61562797c5d18714&version=stable

Reply:
I think it's only a `panic` in debug mode. For the time being, I must admit that I'd prefer it to be a `panic` in release mode too.

Reply:
Fine by me.